### PR TITLE
CMake: MSVC warning C4100: unreferenced formal parameter

### PR DESCRIPTION
**Pull request by vtraag**
This PR addresses warning `C4100: unreferenced formal parameter` mentioned in #1507. Most often I marked this argument as unused, using `IGRAPH_UNUSED`. However, there were also a number of instances for which the arguments really did seem superfluous. In that case, I did remove the actual argument. Please do check if you are OK with this.

There still remain a number of `C4100` warnings that occus in `bliss` and `cs`, but I don't think we should address those, so we should just silence them I think. If you agree I will add the relevant no-warning flags for `bliss` and `cs`.

### CODE REVIEW

File: include/igraph_qsort.h
Comment (ntamas) [inline include/igraph_qsort.h:28]: Do you remember why this include is necessary here? The only thing that is non-standard in this file seems to be `size_t`, which should be covered by `stddef.h`.

File: src/layout.c
Comment (ntamas) [inline src/layout.c:2351]: The "options" argument should be removed IMHO. It used to be used when `igraph_layout_mds` used ARPACK to calculate eigenvectors, but it was a bad decision anyway because the matrices that MDS works on are not sparse. Currently the documentation says that this argument is ununsed, and since we are breaking the API anyway for 0.9, we can remove the argument as I see zero chance that it gets re-introduced in the future.

File: src/other.c
Comment (ntamas) [inline src/other.c:266]: In cases when the arguments are unused but must be there because they are part of a callback signature, I think an easier solution is just to remove the _name_ of the argument and leave its type there. Can you test whether it would work with MSVC? If so, let's do it this way because it improves readability.

File: src/other.c
Comment (ntamas) [inline src/other.c:363]: There's a typo there.

**Review by ntamas — CHANGES_REQUESTED** [review]
A few comments here and there -- see the source. I think the biggest change that I would like to test is whether simply omitting the _name_ of unused arguments from _callback functions_ (where we are forced to add the argument because of the API that consumes the callback) would fix the warnings there. If this is the case, it is a better solution than throwing `IGRAPH_UNUSED()` everywhere.

File: include/igraph_qsort.h
Comment (vtraag) [reply include/igraph_qsort.h:28]: The `IGRAPH_UNUSED` definition is in this header. I would be more than happy to move this definition, because it does not really seem to fit in this header, but that would touch a lot of more files then presumably.

File: src/layout.c
Comment (vtraag) [reply src/layout.c:2351]: I would be fine with removing. The reason that I left it in is because I couldn't foresee whether this argument would be useful at some point when we would like to change the implementation. To me it sounds reasonable that we will never use `ARPACK` for this, so it might make sense to remove it then indeed.

File: include/igraph_qsort.h
Comment (ntamas) [reply include/igraph_qsort.h:28]: Got it. We could move it to `igraph_decls.h` -- after all, it _declares_ that a variable is unused :)

File: src/layout.c
Comment (ntamas) [reply src/layout.c:2351]: Okay, let's remove it then!

File: src/other.c
Comment (vtraag) [reply src/other.c:266]: Unfortunately that does not work. Or well, that is, it gives a different warning: `error C2055: expected formal parameter list, not a type list`. According to the Microsoft docs, this is also expected by ANSI C, so this is then a warning that will presumably also pop up with other compilers. So, probably best to keep this as is.

File: src/other.c
Comment (vtraag) [reply src/other.c]: Fixed.

Comment (vtraag) [reply src/other.c]: Fixed

File: include/igraph_qsort.h
Comment (vtraag) [reply include/igraph_qsort.h:28]: I am now doing this, I just move the declaration to `igraph_decls.h` and remove the `igraph_types.h` inclusion here. It might be that there are some superfluous includes still on `igraph_types.h`, although that is unlikely.

File: include/igraph_adjlist.h
Comment (szhorvat) [inline include/igraph_adjlist.h:190]: If we deprecate `adjedgelist_*` functions (we should), we should not change them anymore.

File: include/igraph_adjlist.h
Comment (szhorvat) [inline include/igraph_adjlist.h:88]: I would not change this, because:

 - Based on our discussion about counting self-loops for eigenvector centrality and other cases, `adjlist_remove_duplicate` and `inclist_remove_duplicate` are not good solutions for what they do. Self-loop counting should be controlled when generating the adjlist/inclist. Therefore, we might soon deprecate this function. Thus, I would not create the impression that it is maintained.
 - The corresponding `adjlist_remove_duplicate` _must_ take a graph argument, consistency is worth maintaining
 - inclists, defined in terms of edge IDs, are not meaningful without a graph

Overall, I would prefer not to touch this function now (and get rid of it soon).

File: include/igraph_embedding.h
Comment (szhorvat) [inline include/igraph_embedding.h:57]: Agreed, remove this!

File: src/scg.c
Comment (szhorvat) [inline src/scg.c:2122]: We should create a wishlist issue to implement the function of this parameter, otherwise it will get forgotten. IMO, this should be done for every similar case (non-functional documented arguments in public functions)

File: include/igraph_adjlist.h
Comment (ntamas) [reply include/igraph_adjlist.h:190]: Yes, that's true, although it doesn't hurt to modify them one last time here. :) I leave it up to Vincent to decide whether he tackles the removal in this PR as well.

File: src/triangles.c
Comment (szhorvat) [inline src/triangles.c:431]: Does anyone fully understand the design of this function? Are we *sure* that there is no reason for, or benefit from having the four different versions of this function have the same signature?

This is one of those things I would not change without a deeper understanding. This change is literally aesthetic to silence a minor compiler warning that is not even on by default.

**Comment by szhorvat** [comment]
That's all, no more comments from me.

File: src/triangles.c
Comment (vtraag) [reply src/triangles.c:431]: This function is explicitly only called in the condition that all vertices are requested. To me that suggests that its intention is to never actually do something with this argument.

Instead, having the same arguments in all functions seems just esthetics to me. But if you find that they are somehow defined elsewhere to have the same call signature, I'd be happy to learn about it.

Comment (ntamas) [reply src/triangles.c:431]: I vaguely remember that Gábor has experimented with several different implementations of the transitivity function and that some of them later turned out to be efficient only if you need the transitivities of (almost) all vertices -- but then we did not bother removing the `vids` argument. There is no underlying deeper design decision here that would necessitate keeping the argument there, especially because these functions are meant to be internal anyway, so if later we somehow realize that they could still be made efficient for a subset of vertices only, we can still re-add the arguments without breaking the API (if we remove them from the public API now).

So, my vote is for removing the arguments and then renaming the functions to `igraph_i_...` and making them static.

Comment (szhorvat) [reply src/triangles.c:431]: Yes, let us remove them from the headers in this PR instead of fussing with the develop branch. I checked and they are not used anywhere, so we will not need to create an internal header for them.

File: src/scg.c
Comment (szhorvat) [reply src/scg.c:2122]: Done: https://github.com/igraph/igraph/issues/1541

File: include/igraph_adjlist.h
Comment (ntamas) [reply include/igraph_adjlist.h:88]: Agreed.

File: src/triangles.c
Comment (vtraag) [reply src/triangles.c:431]: If we remove this in this PR, then why not also remove the `*remove_duplicates` functions in this PR?

Comment (ntamas) [reply src/triangles.c:431]: We should rename the PR in that case to make this explicit (or at least add it in the PR description) -- but otherwise I'm fine with it.

Comment (szhorvat) [reply src/triangles.c:431]: `igraph_transitivity_local_undirected4` was undocumented, and one might argue that it was fairly obvious that it was never meant for public use (i.e. it was exposed mistakenly). Therefore, removing it immediately makes sense.

Functions which were meant to be part of the public API are better deprecated for a version before removing them entirely.

Since all documented functions with `adjedgelist` in the name have been deprecated since 0.6, I support removing `igraph_adjedgelist_remove_duplicate` (even if it did not have an explicit deprecation notice).

----

As for `igraph_inclist_remove_duplicate`, I would not make changes to it in this PR, as there is no replacement functionality for it yet (i.e. add a flag controlling self-loop duplication to `inclist_init`). I think this one needs more discussion.

Comment (szhorvat) [reply src/triangles.c:431]: Anyway, deciding whether we do the removal in this PR or later should definitely not block merging this PR! If there's a doubt, let's leave it for later and go ahead with this PR.

Comment (vtraag) [reply src/triangles.c:431]: OK, let's otherwise clean up those things in separate PRs, that might make it more targeted. I will undo the removal of the argument for now then, and in a next PR I will remove the deprecated `adjedgelist` functions, and in yet another we do something with `inclist_remove_duplicates`.

File: include/igraph_adjlist.h
Comment (vtraag) [reply include/igraph_adjlist.h]: As said in other comments, let's do this in a separate PR.

File: include/igraph_adjlist.h
Comment (vtraag) [reply include/igraph_adjlist.h]: As said in other comments, let's do this in a separate PR. I have now added back in the `graph` argument.

**Comment by vtraag** [comment]
As far as I am concerned we can (squash and) merge this.

**Pull request merged**
